modAionImpl/src/org/aion/zero/impl/sync/BlockPropagationHandler.java

```diff
@@ -0,0 +1,147 @@
+import org.aion.zero.impl.types.AionBlock;
+import org.aion.zero.types.AionTransaction;
+import org.aion.zero.types.IAionBlock;
+import org.apache.commons.collections4.map.LRUMap;
+import org.slf4j.Logger;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Handles state and actions related to block propagation
+ *
+ * TODO: exists as functionality of SyncMgr, need to decouple
+ */
+public class BlockPropagationHandler {
+
+    public enum Status {
+        DROPPED,
+        PROPAGATED,
+        CONNECTED
+    }
+
+    /**
+     * Size of the cache maintained within the map, a lower cacheSize
+     * saves space, but indicates we may "forget" about a block sooner.
+     *
+     * This can possibly lead to increase network traffic and unecessary
+     * process (?)
+     */
+    private final int cacheSize;
+
+    /**
+     * Connection to blockchain
+     */
+    private IAionBlockchain blockchain;
+
+    /**
+     * LRU cache map, maintains the latest cacheSize blocks seen (not counting duplicates).
+     */
+    private final Map<ByteArrayWrapper, Boolean> cacheMap;
+
+    private final IP2pMgr p2pManager;
+
+    private final BlockHeaderValidator blockHeaderValidator;
+
+    private static final Logger log = AionLoggerFactory.getLogger(LogEnum.SYNC.name());
+
+    public BlockPropagationHandler(final int cacheSize,
+                                   final IAionBlockchain blockchain,
+                                   final IP2pMgr p2pManager,
+                                   BlockHeaderValidator headerValidator) {
+        this.cacheSize = cacheSize;
+
+        // guarded by self
+        this.cacheMap = new LRUMap<>(this.cacheSize);
+
+        // the expectation is that we will not have as many peers as we have blocks
+        this.blockchain = blockchain;
+
+        // record our own nodeId to cover corner case
+        this.p2pManager = p2pManager;
+
+        this.blockHeaderValidator = headerValidator;
+    }
+
+    // propagate a new block, send without discrimination to maximize chances
+    public void sendNewBlock(final AionBlock block) {
+        synchronized (this.cacheMap) {
+            this.cacheMap.put(new ByteArrayWrapper(block.getHash()), true);
+        }
+
+        this.p2pManager.getActiveNodes().values().forEach(n -> {
+            if (log.isDebugEnabled())
+                log.debug("sending new block" + block.getShortHash() + " to: " + n.getIdHash());
+            this.p2pManager.send(n.getId(), new BroadcastNewBlock(block));
+        });
+    }
+
+    public Status processIncomingBlock(final byte[] nodeId, final AionBlock block) {
+        if (nodeId == null)
+            return Status.DROPPED;
+
+        if (block == null)
+            return Status.DROPPED;
+
+        ByteArrayWrapper hashWrapped = new ByteArrayWrapper(block.getHash());
+
+        // guarantees if multiple requests of same block appears, only one goes through
+        synchronized(this.cacheMap) {
```

Comment:
Possible to attack cache here?

Flood node with well formed but invalid blocks. Fill the cache with junk causing it to re-send valid but already seen blocks.

Flip order and validate header first before adding to the map?

Reply:
Thanks,  I've moved the check for a well-formed header above the cache. This way at the very least and attack has to go through PoW to invalid the cache.